src/ansi-c/c_typecheck_base.cpp

```diff
@@ -727,6 +727,7 @@ void c_typecheck_baset::typecheck_declaration(
         if(as_const(code_type).requires().is_not_nil())
         {
           auto &requires = code_type.requires();
+          disallow_history_variables(requires);
           typecheck_expr(requires);
           implicit_typecast_bool(requires);
```

Comment:
How about doing that call to `disallow_history_variables` after the typechecking, which would simplify it significantly.